### PR TITLE
QA Release - 7 Dec - 2020

### PR DESCRIPTION
**Pull request by aks30**

### CODE REVIEW

File: migrations/20201127170129-update-user-roles-in-entities-es.js [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    const checkIfIndexExists = await es.indices.exists({ index: entitiesIndex });
+
+    if (checkIfIndexExists.statusCode !== 200) {
+      const createIndex = await es.indices.create({ index: entitiesIndex });
+      if (createIndex.statusCode != 200) {
+        throw new Error("Error while creating entities index.")
+      }
+    }
+
+    let users = await db.collection('userExtension').find({}).project({ _id: 1, roles: 1, userId: 1 }).toArray();
```
Comment (aks30) [inline migrations/20201127170129-update-user-roles-in-entities-es.js:25]: @deepakr123 - Project only _id here then in chunks load the user from userExtension

File: module/entities/helper.js [hunk 2]
```diff
@@ -1338,6 +1338,117 @@ module.exports = class EntitiesHelper {
...
+                    )
+                   
+                    if (entityDocument.statusCode == httpStatusCode.ok.status) {
+
+                        entityDocument = entityDocument.body["_source"].data;
+
+                        if (entityDocument.roles[role.code]) {
+                            if (!entityDocument.roles[role.code].includes(userId)) {
+                                entityDocument.roles[role.code].push(userId);
+                            }
```
Comment (aks30) [inline module/entities/helper.js:1371]: @deepakr123 - We should not update if its not required.

File: module/entityAssessors/helper.js [hunk 3]
```diff
@@ -952,4 +944,61 @@ module.exports = class EntityAssessorHelper {
...
+    /**
+      * Bulk create assessments By entityId and role.
+      * @method
+      * @name bulkCreateByUserRoleAndEntity - Bulk create assessments by entity and role.
+      * @param {Object} userAssessmentData - user assessment data
+      * @param {String} userToken - logged in user token.
+      * @returns {Object}  Bulk create user assessments.
+     */
+
+    static bulkCreateByUserRoleAndEntity(userAssessmentData, userId,userToken) {
```
Comment (aks30) [inline module/entityAssessors/helper.js:958]: @deepakr123 - No need of token here, internal access token only should be required to access the API

File: module/entityAssessors/helper.js [hunk 4]
```diff
@@ -952,4 +944,61 @@ module.exports = class EntityAssessorHelper {
...
+                        "programId" : userAssessmentData.programId,
+                        "solutionId": userAssessmentData.solutionId,
+                        "role": userAssessmentData.assessorRole
+                    })
+                })) 
+        
+                await this.upload
+                (
+                   assessorData,
+                   null,
```
Comment (aks30) [inline module/entityAssessors/helper.js:989]: @deepakr123 - Instead of passing null pass here.

File: module/observations/validator/v1.js [hunk 5]
```diff
@@ -41,6 +41,11 @@ module.exports = (req) => {
         verifyLink : function () {
             req.checkParams('_id').exists().withMessage("required link")
         },
+        bulkCreateByUserRoleAndEntity :  function () {
+            req.checkBody('entityId').exists().withMessage("required entity id");
```
Comment (aks30) [inline module/observations/validator/v1.js:45]: @deepakr123 - please validate mongId

Comment (deepakr123) [reply module/observations/validator/v1.js:45]: resolved

File: migrations/20201127170129-update-user-roles-in-entities-es.js (hunk 1 above)
Comment (deepakr123) [reply migrations/20201127170129-update-user-roles-in-entities-es.js:25]: resolved

File: module/entities/helper.js (hunk 2 above)
Comment (deepakr123) [reply module/entities/helper.js:1371]: resolved

File: module/entityAssessors/helper.js (hunk 3 above)
Comment (deepakr123) [reply module/entityAssessors/helper.js:958]: resolved

**Pull request merged**